### ISSUE TITLE
XAML: MouseBinding does not work if a keyboard modifier is hold down

### ISSUE THREAD

**Issue by Symbai** [dotnet/wpf#6462]
* .NET Core Version: 6.0
* Windows version: Windows 11
* Does the bug reproduce also in WPF for .NET Framework 4.8?: No idea
 
 **Problem description:**
Binding a command in XAML using MouseBinding is not executed when the user holds down SHIFT or CTRL key while clicking on it. This is a problem if we want to check for these modifier keys and execute different code in our command based on whether SHIFT or CTRL or no modifier key is hold down
 
 **Actual behavior:**
The command is not executed
 
 **Expected behavior:**
The command should be executed. For reference a KeyBinding for example is executed even if a mouse button is hold down. Mouse and Key bindings should work independently.
 
 **Minimal repro:**
```XAML
<Border Width="400" Height="400" Background="Red">
	<Border.InputBindings>
		<MouseBinding MouseAction="LeftClick" Command="{Binding MyCommand}"/>
	</Border.InputBindings>
</Border>
```
```C#
public RelayCommand MyCommand => new(a =>
{
	if (Keyboard.IsKeyDown(Key.LeftCtrl))
		MessageBox.Show("Hey");
	else if (Keyboard.IsKeyDown(Key.LeftShift))
		MessageBox.Show("Nay");
	else
		MessageBox.Show("Okay");
});
```

**Comment by miloush** [comment]
Note the notes for the Gesture property you linked:

> When defining a [MouseBinding](https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.mousebinding?view=windowsdesktop-6.0) in Extensible Application Markup Language (XAML), there are two ways to specify the [MouseGesture](https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.mousegesture?view=windowsdesktop-6.0). The first way to establish a [MouseBinding](https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.mousebinding?view=windowsdesktop-6.0) in XAML is to define the [Gesture](https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.mousebinding.gesture?view=windowsdesktop-6.0) attribute of the [MouseBinding](https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.mousebinding?view=windowsdesktop-6.0) element, which enables a syntax to specify mouse actions and modifiers as a single string; for example, "CTRL+LeftClick". The second way is to define the [MouseAction](https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.mousebinding.mouseaction?view=windowsdesktop-6.0) attribute of the [MouseBinding](https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.mousebinding?view=windowsdesktop-6.0) element. **Both ways of setting the [MouseGesture](https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.mousegesture?view=windowsdesktop-6.0) are equivalent** and modify the same underlying object, but there will be a conflict if both are used. In the case when the [MouseAction](https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.mousebinding.mouseaction?view=windowsdesktop-6.0) and the [Gesture](https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.mousebinding.gesture?view=windowsdesktop-6.0) attributes are both set, the gesture is invalid. In general, it is recommended that you use only the [Gesture](https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.keybinding.gesture?view=windowsdesktop-6.0) attribute from XAML, even if you do not specify modifiers; this avoids ambiguity, provides the most streamlined syntax, and provides the most straightforward representation for serialization.

If the behavior was changed the way you suggest, this contract would be broken - using `MouseAction="LeftClick"` would not be equivalent to using `Gesture="LeftClick"`. You can obviously argue that the `Gesture` behaviour in such case should also be changed to mean any modifier. That would however make `KeyBinding.Gesture` and `MouseBinding.Gesture` behave differently.

Either way, you would be breaking existing code. Commands would be invoked in unexpected situations (i.e. with modifiers present), or, since only the first matching binding is executed, commands could stop being executed.

Another way that one binding per multiple combinations of modifiers could be achieved would be to add a property on `MouseGesture` and possibly `KeyGesture` that would allow such behavior. This could be a simple boolean property, or something like `OptionalModifiers`. The problem again is that this would need to be serializable and any new syntax introduced would break existing code not expecting it.

There does not seem to be an easy compatible way out and I am not quite sure the benefits are worth the costs here.

**Comment by Symbai** [comment]
> Either way, you would be breaking existing code. Commands would be invoked in unexpected situations (i.e. with modifiers present), or, since only the first matching binding is executed, commands could stop being executed.

Not really. If Gesture is used Gesture will be (should be) checked first. The only reason to use Gesture is when you explicit want to execute a different(!) command if a modifier is hold down. If Gesture isn't used or the gesture doesn't match the modifier MouseBinding will be executed.

Its odd that if I want to execute the same command regardless of the modifier I have to add 3(!) Gesture's in XAML all pointing to the same command.

**Comment by miloush** [comment]
Take this existing code:
```xml
<MouseBinding MouseAction="LeftClick" Command="{Binding MyCommand1}" />
<MouseBinding Gesture="Control+LeftClick" Command="{Binding MyCommand2}" />
```
If what you propose is implemented,
 a) `MyCommand1` is now unexpectedly called when Ctrl is down
 b) `MyCommand2` stops being invoked completely

> I want to execute the same command regardless of the modifier I have to add 3(!) Gesture's

This is one of the dangers of the design - it is not true. You have to add all combinations of all the modifiers, i.e. 16 bindings (if we only consider the defined `ModifierKeys` values) to get an equivalent behavior to what you propose.

However, I don't see a reason why you couldn't create your own `MouseOnlyBinding` that would only check for the mouse buttons?
